### ISSUE TITLE
java.lang.ClassNotFoundException: java.util.HashMap$Values

### ISSUE THREAD

**Issue by molok** [streamsupport/streamsupport#13]
After deploying a web application on Websphere 8.5.0.1 (which runs on IBM JVM) the webserver throws a java.lang.ClassNotFoundException.
Could it be a different implementation of java.util.HashMap in IBM JVM as opposed to OpenJDK and Oracle?

Environment
```
WebSphere [BPMPC 8.5.0.1 20131107-140634]Platform 8.5.5.1 [ND 8.5.5.1 cf011341.03] running with process name PCCell1\Node1\SingleClusterMember1 and process id 13369756
Host Operating System is AIX, version 6.1
Java version = 1.6.0, Java Compiler = j9jit26, Java VM name = IBM J9 VM

```

my code:
```
HashMap<String, String> configParams = new HashMap<>();
/* more code */
Map<String, AuthEndpoint> newFormat = StreamSupport.stream(configParams.entrySet())
        .filter(e -> e.getKey().endsWith("_uri"))
        .collect(Collectors.toMap(
                e -> e.getKey().substring(0, e.getKey().indexOf("_uri")),
                e -> new AuthEndpoint(uri(e.getValue()))));
```

stacktrace:
```
0000018f webapp        E com.ibm.ws.webcontainer.webapp.WebApp notifyServletContextCreated SRVE0283E: Exception caught while initializing context: {0}
                                 java.lang.Error: java.lang.ClassNotFoundException: java.util.HashMap$Values
        at java8.util.HMSpliterators.<clinit>(HMSpliterators.java:451)
        at java.lang.J9VMInternals.initializeImpl(Native Method)
        at java.lang.J9VMInternals.initialize(J9VMInternals.java:236)
        at java8.util.Spliterators.setSpliterator(Spliterators.java:1001)
        at java8.util.Spliterators.spliterator(Spliterators.java:931)
        at java8.util.stream.StreamSupport.stream(StreamSupport.java:120)
        at my.app.config.ApplicationConfig$RealApplicationConfig.oldFormatHosts(ApplicationConfig.java:125)
        at my.app.config.ApplicationConfig$RealApplicationConfig.getAllEndpoints(ApplicationConfig.java:103)
        at my.app.config.ApplicationConfig$RealApplicationConfig.getEndpointByName(ApplicationConfig.java:116)
        at my.app.app.ApplicationBuilder.XDirectoryPoller(ApplicationBuilder.java:34)
        at my.app.web.DirectoryListener.contextInitialized(DirectoryListener.java:22)
        at com.ibm.ws.webcontainer.webapp.WebApp.notifyServletContextCreated(WebApp.java:1682)
        at com.ibm.ws.webcontainer.webapp.WebAppImpl.initialize(WebAppImpl.java:414)
        at com.ibm.ws.webcontainer.webapp.WebGroupImpl.addWebApplication(WebGroupImpl.java:88)
        at com.ibm.ws.webcontainer.VirtualHostImpl.addWebApplication(VirtualHostImpl.java:169)
        at com.ibm.ws.webcontainer.WSWebContainer.addWebApp(WSWebContainer.java:749)
        at com.ibm.ws.webcontainer.WSWebContainer.addWebApplication(WSWebContainer.java:634)
        at com.ibm.ws.webcontainer.component.WebContainerImpl.install(WebContainerImpl.java:426)
        at com.ibm.ws.webcontainer.component.WebContainerImpl.start(WebContainerImpl.java:718)
        at com.ibm.ws.runtime.component.ApplicationMgrImpl.start(ApplicationMgrImpl.java:1175)
        at com.ibm.ws.runtime.component.DeployedApplicationImpl.fireDeployedObjectStart(DeployedApplicationImpl.java:1370)
        at com.ibm.ws.runtime.component.DeployedModuleImpl.start(DeployedModuleImpl.java:639)
        at com.ibm.ws.runtime.component.DeployedApplicationImpl.start(DeployedApplicationImpl.java:968)
        at com.ibm.ws.runtime.component.ApplicationMgrImpl.startApplication(ApplicationMgrImpl.java:774)
        at com.ibm.ws.runtime.component.ApplicationMgrImpl.startApplicationDynamically(ApplicationMgrImpl.java:1374)
        at com.ibm.ws.runtime.component.ApplicationMgrImpl.start(ApplicationMgrImpl.java:2179)
        at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.start(CompositionUnitMgrImpl.java:445)
        at com.ibm.ws.runtime.component.CompositionUnitImpl.start(CompositionUnitImpl.java:123)
        at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.start(CompositionUnitMgrImpl.java:388)
        at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.access$500(CompositionUnitMgrImpl.java:116)
        at com.ibm.ws.runtime.component.CompositionUnitMgrImpl$1.run(CompositionUnitMgrImpl.java:663)
        at com.ibm.ws.security.auth.ContextManagerImpl.runAs(ContextManagerImpl.java:5474)
        at com.ibm.ws.security.auth.ContextManagerImpl.runAsSystem(ContextManagerImpl.java:5600)
        at com.ibm.ws.security.core.SecurityContext.runAsSystem(SecurityContext.java:255)
        at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.startCompositionUnit(CompositionUnitMgrImpl.java:677)
        at com.ibm.ws.runtime.component.CompositionUnitMgrImpl.startCompositionUnit(CompositionUnitMgrImpl.java:621)
        at com.ibm.ws.runtime.component.ApplicationMgrImpl.startApplication(ApplicationMgrImpl.java:1266)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:60)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:37)
        at java.lang.reflect.Method.invoke(Method.java:611)
        at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:69)
        at sun.reflect.GeneratedMethodAccessor100.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:37)
        at java.lang.reflect.Method.invoke(Method.java:611)
        at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:272)
        at javax.management.modelmbean.RequiredModelMBean$4.run(RequiredModelMBean.java:1148)
        at java.security.AccessController.doPrivileged(AccessController.java:298)
  at com.ibm.oti.security.CheckedAccessControlContext.securityCheck(CheckedAccessControlContext.java:30)
        at sun.misc.JavaSecurityAccessWrapper.doIntersectionPrivilege(JavaSecurityAccessWrapper.java:41)
        at javax.management.modelmbean.RequiredModelMBean.invokeMethod(RequiredModelMBean.java:1142)
        at javax.management.modelmbean.RequiredModelMBean.invoke(RequiredModelMBean.java:995)
        at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:847)
        at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:783)
        at com.ibm.ws.management.AdminServiceImpl$1.run(AdminServiceImpl.java:1335)
        at com.ibm.ws.security.util.AccessController.doPrivileged(AccessController.java:118)
        at com.ibm.ws.management.AdminServiceImpl.invoke(AdminServiceImpl.java:1228)
        at com.ibm.ws.management.connector.AdminServiceDelegator.invoke(AdminServiceDelegator.java:181)
        at com.ibm.ws.management.connector.ipc.CallRouter.route(CallRouter.java:247)
        at com.ibm.ws.management.connector.ipc.IPCConnectorInboundLink.doWork(IPCConnectorInboundLink.java:360)
        at com.ibm.ws.management.connector.ipc.IPCConnectorInboundLink$IPCConnectorReadCallback.complete(IPCConnectorInboundLink.java:602)
        at com.ibm.ws.ssl.channel.impl.SSLReadServiceContext$SSLReadCompletedCallback.complete(SSLReadServiceContext.java:1784)
        at com.ibm.ws.tcp.channel.impl.AioReadCompletionListener.futureCompleted(AioReadCompletionListener.java:175)
        at com.ibm.io.async.AbstractAsyncFuture.invokeCallback(AbstractAsyncFuture.java:217)
        at com.ibm.io.async.AsyncChannelFuture.fireCompletionActions(AsyncChannelFuture.java:161)
        at com.ibm.io.async.AsyncFuture.completed(AsyncFuture.java:138)
        at com.ibm.io.async.ResultHandler.complete(ResultHandler.java:204)
        at com.ibm.io.async.ResultHandler.runEventProcessingLoop(ResultHandler.java:816)
        at com.ibm.io.async.ResultHandler$2.run(ResultHandler.java:905)
        at com.ibm.ws.util.ThreadPool$Worker.run(ThreadPool.java:1862)
Caused by: java.lang.ClassNotFoundException: java.util.HashMap$Values
        at java.lang.Class.forNameImpl(Native Method)
        at java.lang.Class.forName(Class.java:177)
        at java8.util.HMSpliterators.<clinit>(HMSpliterators.java:439)
        ... 69 more
```

**Comment by molok** [comment]
I can confirm that running this snippet on IBM JVM throws a ClassNotFoundException:

```
public class Main {
    public static void main(String[] args) throws Exception {
        System.out.println("before");
        Class.forName("java.util.HashMap$Values");
        System.out.println("after");
    }
}
```

```
$ /opt/IBM/BPM/v8.5/ProcCen/java/jre/bin/java -version
java version "1.6.0"
Java(TM) SE Runtime Environment (build pap6460_26sr6ifix-20130824_01(SR6+IV44889+IX90121+IX90123+IV45933+IV47046+IV45369+IV47468))
IBM J9 VM (build 2.6, JRE 1.6.0 AIX ppc64-64 Compressed References 20130823_162690 (JIT enabled, AOT enabled)
J9VM - R26_Java626_SR6_Ifix_20130823_2006_B162690
JIT  - r11.b04_20130528_38954ifx6
GC   - R26_Java626_SR6_Ifix_20130823_2006_B162690_CMPRSS
J9CL - 20130823_162690)
JCL  - 20130618_01

```

**Comment by stefan-zobel** [comment]
Hi Alessio, @molok

IBM Java VMs are not officially supported. Wich Java version (6, 7, 8) do you use in your WebSphere environment?

However, not everything is lost.
Your best bet is setting a VM system property from your code **before** any class from the streamsupport.jar gets loaded:


`System.setProperty("java8.util.Spliterators.assume.oracle.collections.impl", "false");`

 
(or try the same with the "-D" VM switch).

Please let me know if this solves your problem.

Kind regards,
Stefan

**Comment by molok** [comment]
Hi @stefan-zobel ,
we are still stuck on the IBM Java 6 JVM (the one I posted before)
Settings the system properties DOES fix the issue, I'm very grateful for your quick support and this awesome library that eases the pain of Java 6.
Is there anything else I should know about streamsupport and IBM JVM?
I see this propery was documented before in the README.md but was removed afterwards.

Do you think the library could have a fallback mechanism to handle cases like this more gracefully?

**Comment by stefan-zobel** [comment]
Hi @molok,

well, this switch **is** the fallback for non-OpenJDK based VMs.

It was introduced in release 1.1.3 (ticket 12, https://sourceforge.net/p/streamsupport/tickets/12/) mainly as a preliminary workaround for Android which, nowadays, has its own specialized implementation. However, I also did run the the test suite on IBM Java 6, 7 and 8 every now and then (even if I don't support IBM VMs formally).

You are right, I did remove the explicit reference to IBM in the last release (maybe a bit too overhasty). Though a more implicit reference is still in the readme.txt (https://github.com/streamsupport/streamsupport/blob/master/Readme.txt) in the "1.1.3-stable" section (search for "non-OpenJDK based"). I'll add back a corresponding note in the upcoming release.

As for specifics of streamsupport on the IBM VM: my tests at that time indicated that everything works fine on the Java 8 VM (that is, because IBM licensed the Oracle implementation). There were a few test failures (about 5 - 8 failures out of ~ 15000 tests I had at that time) on Java 6 and Java 7 that are caused by bugs in the IBM collection API implementation. I don't remember exactly, but I think the problems were associated with j.u.LinkedHashMap and j.u.IdentityHashMap and very uncommon access patterns of these collections in the test suite.

In short: I believe it is very unlikely that you'll experience further problems if you set that switch. However, I haven't run the test suite on IBM VMs for a very long time. So it's theoretically possible that I've destroyed something in more recent releases that used to work earlier.

Please feel free to create a new ticket if you run across further problems. In that case, I'll see if I can do something in an upcoming release (but please understand that IBM VMs will remain unsupported "officially").

Kind regards,
Stefan

**Comment by stefan-zobel** [comment]
Hi @molok,

I've indeed found a new bug (introduced in 1.5.2) associated with the

`java8.util.Spliterators.assume.oracle.collections.impl=false`

switch.

Please don't use **parallel** streams generated from `j.u.c.CopyOnWriteArrayList` with this switch (sequential streams are safe).

I've filed a ticket (https://sourceforge.net/p/streamsupport/tickets/300/). This problem will be resolved in the upcoming 1.5.5 release.

I'm also planning to re-run the test suite against the current 1.5.5-SNAPSHOT on IBM Java 6 & 7. I'll keep you updated if I find any new problems.

Regards,
Stefan

**Comment by stefan-zobel** [comment]
@molok

IBM Java 6 & 7 re-test results: Please stay away from release 1.5.4 if you want to do anything with **parallel** streams on IBM Java 6 or 7!

I've botched that up completely in the latest release. In this case, instead use **release 1.5.3**.

I've filed a ticket (https://sourceforge.net/p/streamsupport/tickets/301/). This problem will be resolved in the upcoming 1.5.5 release.

Other than that: IBM's `java.util.WeakHashMap` and `java.util.IdentityHashMap` are problematic (don't use them with streams), `java.util.LinkedHashMap` mostly works (3 test failures with quite uncommon access patterns).

Everything else  (insofar it is covered by the test suite) works fine in the current 1.5.5-SNAPSHOT.

**Comment by stefan-zobel** [comment]
@molok

streamsupport release 1.5.5 got released today. Please use that version on IBM JRE 6.

Kind regards,
Stefan

**Comment by molok** [comment]
Thank you very much for your outstanding support

Il 25 mag 2017 5:25 PM, "stefan-zobel" <notifications@github.com> ha
scritto:
